commit 67756872bd6f9f4e58776d3951adaea93328a8ba
Author: Dimitrios Apostolou <dimitrios.apostolou@cfengine.com>
Date:   Sat Dec 20 12:07:03 2014 +0100

    Redmine#6672: Memory exhausted with list larger than 4994 items
    
    Convert Yacc right recursion to left recursion so that stack space is
    not being exhausted when parsing long lists.
    (cherry picked from commit a28b28e93335bd9907e91739af401061757f5a2f)

diff --git a/libpromises/cf3parse.y b/libpromises/cf3parse.y
index 1bd37da..da514c3 100644
--- a/libpromises/cf3parse.y
+++ b/libpromises/cf3parse.y
@@ -960,13 +960,15 @@ rval:                  IDSYNTAX
 
 /* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
 
-list:                  OB litems CB
+list:                  OB CB
+                     | OB litems CB
+                     | OB litems ',' CB
 
 /* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
 
-litems:                /* empty */
-                     | litem
-                     | litem ','  litems
+litems:
+                       litem
+                     | litems ',' litem
                      | litem error
                        {
                            ParserDebug("P:rval:list:error yychar = %d\n", yychar);
